### ISSUE TITLE
Need help debugging crashing of zsh with zplugin and some prezto modules

### ISSUE THREAD

**Issue by ngocphamm** [zdharma/zplugin#130]
**Describe the bug**
Loading of a few prezto modules results in crashing zsh session (shortly) right after starting up

**To Reproduce**
1. Use the provided configuration
2. Uncomment 2 lines for using prezto module `fasd`
3. Open up a new iTerm2 tab (this is just for quick/easier testing, it happens to me in both Alacritty and iTerm2) and hit `Enter` key, fast, multiple times
4. Zsh tab/session will very likely crash crashes (iTerm2 has a broken pipe popup message)

**Expected behavior**
No crash

**Versions:**
 - $ZSH_VERSION == 5.7.1
 - Operating system name and version (uname -a): macOS 10.14.5

**Additional context**
This is the `.zshrc` I have. There's actually a lot more on top of these lines, but I tried commenting those out and the crash still happens, so I think the issue is within those lines.

```
setopt promptsubst
autoload -Uz promptinit && promptinit

# zstyle ':prezto:*:*' color 'yes'
# zstyle ':prezto:module:editor' key-bindings 'emacs'
# zstyle ':prezto:module:history-substring-search:color' found 'bg=green,fg=white,bold'
# zstyle ":prezto:module:enhancd" command "fzy:fzf:peco"
# zstyle ":prezto:module:enhancd" command "cd"
# # zstyle ':prezto:module:prompt' theme 'sorin'

# ##### Zplugin
source '/Users/ngoc/.zplugin/bin/zplugin.zsh'
autoload -Uz _zplugin
(( ${+_comps} )) && _comps[zplugin]=_zplugin

zplugin light mafredri/zsh-async
zplugin light zsh-users/zsh-completions
zplugin light zsh-users/zsh-history-substring-search
zplugin light zsh-users/zsh-autosuggestions


zplugin ice svn; zplugin snippet PZT::modules/git
zplugin snippet PZT::modules/history/init.zsh

# zplugin ice wait"0" pick"init.zsh" svn
# zplugin snippet PZT::modules/fasd

# zplugin ice wait"0" blockf pick"init.zsh" svn \
#   atclone'git clone https://github.com/zsh-users/zsh-completions.git external'
# zplugin snippet PZT::modules/completion

# zplugin ice wait"1" pick"init.zsh" svn \
#   atclone'git clone https://github.com/zsh-users/zsh-history-substring-search.git external'
# zplugin snippet PZT::modules/history-substring-search

# zplugin ice wait"2" pick"init.zsh" svn \
#   atclone'git clone https://github.com/zsh-users/zsh-autosuggestions.git external'
# zplugin snippet PZT::modules/autosuggestions

# # zplugin ice as"program" pick"prompt-pwd" silent svn; zplugin snippet PZT::modules/prompt
# # zplugin ice svn silent; zplugin snippet PZT::modules/prompt

zplugin ice wait"0" atinit"zpcompinit; zpcdreplay"
zplugin light zdharma/fast-syntax-highlighting
```

The more prezto module lines I uncomment (and comment out the corresponding `zsh-user` plugins), the more often I can reproduce the crash. I know those prezto modules actually use the `zsh-users` repos behind the scene, but I'm currently using prezto and is used to the way it configure things so I just want to start with loading prezto modules first.

Any help would be really appreciated!

Thanks!

**Comment by ngocphamm** [comment]
I think it's somehow the Turbo mode with `fasd`. As soon as I remove `wait"0" pick"init.zsh" from the below, it's fine.

```
zplugin ice wait"0" pick"init.zsh" svn
zplugin snippet PZT::modules/fasd
```

**Comment by robobenklein** [comment]
Perhaps you could try starting zsh in verbose mode (`set -v` in a testing zshrc) and try to see what lines get parsed right before the crash? (Maybe run it in a subshell so you can keep the scrollback output from -v)

**Comment by ngocphamm** [comment]
Just tried it @robobenklein. Weird... It crashed the first time I started a new ZSH tab (iTerm2), and not for the following several times, even though I'm pretty sure it reached the same point as in the gist below (where it did crash/exit)

Then I went to edit the `.zshrc` for a minute or so, came back and started a new set of tabs again. First tab crashed again, and not subsequent tabs. Same point reached (that's probably the end of it too because it lookes like the `fast-zsh-

https://gist.github.com/ngocphamm/dd0af29ad2bb99a7846ca53385ca3804

**Comment by robobenklein** [comment]
Not having subsequent instances crash makes me very suspicious of some kind of caching mechanism gone wrong.

Are you able to reproduce the issue on a clean system? Something like a docker container would work as well... (tis why I put my configs in one for testing: https://github.com/robobenklein/configs/blob/master/Dockerfile )

**Comment by ngocphamm** [comment]
It could be some caching like you said, even though it did have at least one of those subsequent instance crashes, depends on how many I would spin up (new iTerm2 tabs). I probably didn't open enough to see it happen very often, like when I first reported this issue. Basically to me it sounds more random than having a pattern.

I'm not very familiar with Docker. Do you mean to bring up just an unix-based clean system to test? If yes can you help me with some easy to follow guides or something? For a clean mac system, I have no idea if that's even possible without a lot of messing around.

**Comment by robobenklein** [comment]
If you've never used docker, that's probably not an easy path.

Before that, I'd try to get rid of as much cache as you know of, like the zcompdump, try purging the ~/.zplugin folders, etc.

Other ideas could be:
 - Walking over to another unix machine
 - Creating a new user on your machine and using that, then delete it after
 - Run some linux off a USB stick (though Mac + Linux generally has problems)

If you have your configs public I can try running them to see if I get the same problems.

**Comment by laggardkernel** [comment]
I'm basically using all the plugins included in your example on macOS 10.13 and all plugins work well. The conf here extract from my `.zshrc` may help you.

```zsh
# PZT::modules/completion {{{
_pzt_completion_pre() {
  # run under folder PZT::modules/completion, stop auto compinit
  sed -i "/autoload -Uz compinit/,/unset _comp_files/d" ./init.zsh
  sed -i "/fpath/d" ./init.zsh
  # recompile the modified file
  for i in ./*.zsh; do zcompile "$i"; done
}

zplugin ice svn blockf \
atclone'_pzt_completion_pre' \
atpull'_pzt_completion_pre'
zplugin snippet PZT::modules/completion

# }}}
# zsh-users/zsh-completions {{{
# delay compinit
zplugin ice blockf \
atpull'zplg creinstall zsh-users/zsh-completions'
zplugin light zsh-users/zsh-completions

# }}}
# history-search-multi-word {{{
# zstyle :plugin:history-search-multi-word reset-prompt-protect 1
zplugin ice compile"(hsmw-*|history-*)" #wait'0' lucid
zplugin light zdharma/history-search-multi-word # bindkey "^R" could be overridden by PMZ::editor
bindkey "^R" history-search-multi-word

# }}}
# autosuggestions {{{
# ZSH_AUTOSUGGEST_USE_ASYNC=1

zplugin ice svn \
atclone'git clone https://github.com/zsh-users/zsh-autosuggestions external; for i in external/*.zsh; do zcompile $i; done;' \
atpull'cd ./external;git reset --hard HEAD;git pull;for i in ./*.zsh; do zcompile $i; done;' #wait'0' atload"_zsh_autosuggest_start" lucid
zplugin snippet PZT::modules/autosuggestions

# }}}
# 0 fast-syntax-highlighting {{{
zplugin ice wait'0' lucid
zplugin light zdharma/fast-syntax-highlighting

# }}}
# 0 history-substring-search {{{
zplugin ice svn \
atclone'git clone https://github.com/zsh-users/zsh-history-substring-search external; for i in external/*.zsh; do zcompile $i; done;' \
atpull'cd ./external;git reset --hard HEAD;git pull;for i in ./*.zsh; do zcompile $i; done;' wait'0' lucid
zplugin snippet PZT::modules/history-substring-search

# }}}
# compinit params {{{
_compinit_param() {
  # borrowed from prezto
  local _comp_files=(${ZPLGM[ZCOMPDUMP_PATH]:-${ZDOTDIR:-$HOME}/.zcompdump}(Nm-20))
  if (( $#_comp_files )); then
    ZPLGM[COMPINIT_OPTS]="-i -C"
  else
    ZPLGM[COMPINIT_OPTS]="-i"
  fi
}

# }}}
# 0 fasd + completion init {{{
# load after compint to setup command,word completion

zplugin ice svn \
atclone'git clone https://github.com/clvv/fasd external && zcompile external/fasd' \
atpull'command rm -rf external;git clone https://github.com/clvv/fasd external && zcompile external/fasd' \
atinit'_compinit_param;zpcompinit;zpcdreplay;' wait'0' lucid
zplugin snippet PZT::modules/fasd

# }}}
```

Here's some tips for any prezto immigrant.

Firstly, there's no need to load every plugin in turbo mode.

Loading order of prezto plugins matters. Some plugins depend on other plugins.
- `utility > completion > syntax-highlighting > history-substring-search > autosuggestions`
- `utility > completion > syntax-highlighting > prompt`
- `utility > completion > fasd`

Although according to the `README.md` in prezto's repo, `autosuggestions` should be loaded at the end. But after reading from `zsh-users/zsh-autosuggestions`, it seems there's no reason to do that.

See more detail at https://github.com/sorin-ionescu/prezto/issues/1659.

The `PZT::modules/completion` plugin consists of `zsh-users/completions` and some completion related confs. `compinit` is done by `PZT::modules/completion` automatically, you'd better disable it using `sed` (just like what I do), or don't use it at all. Try to fake the loading status of it by,

```zsh
zstyle ":prezto:module:completion" loaded 'yes'
```

**Comment by laggardkernel** [comment]
Your problem may be related to the loading sequence. Except my own config, the modified version of your conf below works on my side as well.

```zsh
zstyle ":prezto:module:completion" loaded 'yes'


zplugin ice wait"0" lucid
zplugin light zdharma/fast-syntax-highlighting

zplugin ice wait"0" pick"init.zsh" svn \
  atclone'git clone https://github.com/zsh-users/zsh-autosuggestions.git external' lucid
zplugin snippet PZT::modules/autosuggestions

zplugin ice wait"0" pick"init.zsh" svn \
  atclone'git clone https://github.com/zsh-users/zsh-history-substring-search.git external' lucid
zplugin snippet PZT::modules/history-substring-search


# zplugin ice as"program" pick"prompt-pwd" silent svn; zplugin snippet PZT::modules/prompt
# zplugin ice svn silent; zplugin snippet PZT::modules/prompt

zplugin ice wait"0" pick"init.zsh" svn lucid atinit"zpcompinit; zpcdreplay"
zplugin snippet PZT::modules/fasd
```

**Comment by ngocphamm** [comment]
Thank you for investing your time on this issue of mine @laggardkernel! I appreciate it very much!

I will be able to try maybe later today, but for now I have a few thoughts.

1. I think it might be the loading sequence like you said. I was aware of the order of `completion`, `history-substring-search`, and `autosuggestions` per README by such Prezto modules, but I forgot about `fasd`.
> utility > completion > syntax-highlighting > history-substring-search > autosuggestions
2. I might be confused with Zplugin's Turbo mode, but I thought having `wait"X"` basically means loads the plugin at the very end? (like after the `completion`)
3. Also, Zplugin's documentation (admittedly quite confusing to me) says `fast-syntax-highlighting` (and any other syntax highlighting plugins) should be loaded last? If that's true shouldn't we do `wait"0"` for `fasd` and `wait"1"` for `fast-syntax-highlighting`?
4. I'm confused again with Zplugin's said ability to "catch" other plugin's `compdefs` and run them later with `zpcompinit; zpcdreplay`. With that said, I thought the `completion` module can be loaded at the top without triggering `compinit`, which is time consuming? Please bear with me because I'm totally not having a deep understanding of how these things work in ZSH (or any other shells, to be honest).
5. Last, the config I posted in my first post (with `fasd` lines uncommented) was actually causing shell crash/exit for me, and I didn't even use any Prezto plugins other than `git`, `history`, and `fasd`, in that order. And `zsh-users/zsh-completions` is "defined" before `fasd` (now that I'm confused with what is actually load order). So looking like this with all the commented lines removed. UNLESS, Zplugin caches the load somewhere... I did check `zplg times` (when the shell didn't crash) and it showed only the modules I didn't comment out, though.

```
zplugin light mafredri/zsh-async
zplugin light zsh-users/zsh-completions
zplugin light zsh-users/zsh-history-substring-search
zplugin light zsh-users/zsh-autosuggestions

zplugin ice svn; zplugin snippet PZT::modules/git
zplugin snippet PZT::modules/history/init.zsh

zplugin ice wait"0" pick"init.zsh" svn
zplugin snippet PZT::modules/fasd

zplugin ice wait"0" atinit"zpcompinit; zpcdreplay"
zplugin light zdharma/fast-syntax-highlighting
```

**Comment by ngocphamm** [comment]
@robobenklein I appreciate your offer to debug this with/for me. As said in the above post, it's the actual config I posted in first post (repeated in the above post with commented lines removed). Of course I have much more in my `.zshrc`, but I tested with everything else, except for what I posted, commented out and the crash/exit still happened.

I'm not familiar with Docker, but I am with Virtual machine and Linux in general, so I can definitely try with some Linux machines later.

So BTW, what's the proper way to "clear the caches" of Zplugin? Just remove the whole `.zplugin` folder and start again with the installer script, or just everything except for `bin` inside `.zplugin`?

**Comment by laggardkernel** [comment]
According to the questions you asked me, especially the question No.2, you lack some basic knowledge about zplugin. Besides, some questions are hard to explain, I still don't know enough about the loading sequence.

1. I'll try to make the explanation about loading sequence more detail. The loading status is detected in prezto by the `zstyle` value `":prezto:module:$pmodule"` they set,

```
# https://github.com/sorin-ionescu/prezto/blob/1f4601e44c989b90dc7314b151891fa60a101251/init.zsh#L95-L98

  # Load Prezto modules.
  for pmodule in "$pmodules[@]"; do
    if zstyle -t ":prezto:module:$pmodule" loaded 'yes' 'no'; then
      continue
    else
```

In fact, `completion` in `utility > completion` sequence may indicate  the **`compinit` process**, not the module itself. And some of part of the loading sequence in prezto **makes no sense at all**.

It seems there's NO any intersection between features of `PZT::modules/completion` and `PZT::modules/syntax-highlighting`. There's totally no reason to load `completion > syntax-highlighting`. I guess that, prezto tells users this just to make sure these two modules are not loaded at the very beginning, or before some other modules. This is the same reason zdharma, the author of zplugin, tells you `fast-syntax-highlighting` should be loaded at the end.

2. Remember this: **`zplugin` is not `zplug`, it doesn't load plugins parallelly.** If you want one plugin be loaded after another, just put the config of the it behind the other. `wait'X'` means turbo mode, which i personally prefer calling it the **background loading**. Normally, plugins in `.zshrc` are loaded before you can type any commands in terminal, plugins in the turbo mode (background) are loaded after the prompt is show, after you can type commands in the terminal. `wait'0'`

BTW: zplug's (not zplugin) parallel loaiding means you have no definite control of the loading order at all. Even plugins with the same `prefer` value are loaded parallelly in zplug.

3. Firstly, different plugins can be configured with same `wait'x'` time, let's say `wait'0'`. They are loaded sequentially in the background (turbo mode). If there're too many plugins set with`wait'0'` and total loading time for `wait'0'` plugins exceeds 1 second, I **guess** the `wait'1'` plugins will be deferred until all the `wait'0'` plugins are loaded. (@psprint the author must know this better than me.)

Secondly, the loading order lists given you before by me doesn't load `syntax-highlighting` at the end. If you read carefully in my config, I don't load `zdharma/fast-syntax-highlighting` ( a optimized version of `zsh-users/zsh-syntax-highlighting` at the end either. So why? 

[Why must zsh-syntax-highlighting.zsh be sourced at the end of the .zshrc file?](https://github.com/zsh-users/zsh-syntax-highlighting#why-must-zsh-syntax-highlightingzsh-be-sourced-at-the-end-of-the-zshrc-file)
> `zsh-syntax-highlighting.zsh` **wraps ZLE widgets**. It must be sourced after all custom widgets have been created (i.e., after all `zle -N` calls and after running `compinit`). Widgets created later will work, but will not update the syntax highlighting.

As far as I know, there're basically 2 reasons why one plugin is loaded after another. 1) It depends on the settings defined in the other plugin. 2) The plugin may broke things set before, so it has to be loaded later.

`PZT::modules/history-substring-search` (a wrapper of `zsh-users/zsh-syntax-highlighting`) and `zdharma/fast-syntax-highlighting` wraps `zle -N`, they don't need to be loaded at the exact end of the conf file. Making sure they're loaded before any other plugins touch the `zle` widgets is enough. You may guessed it, `fasd` doesn't touch `zle` widget at all.

Which plugin is loaded at the very end doesn't matter. Just to make sure `compinit` is done after all pugins containing completions are loaded. `syntax-highlighting` is loaded after all plugins using `zle` widgets.

4. I'm not a professional of zsh, the answer for this question is based on my user experience. 

There're are 2 ways to define a completion. 1) define it as a `autoload` function, which means put the completion function in a separate file (name begins with an under score) under a directory included by `fpath`. Once `compinit` is executed, it will search all the entries in `fpath` to recognize these completion files. 2) define a normal function as a completion function with `compdef`, the problem here is that `compdef` is accessible only after `compinit` is done. (Maybe `compinit` defines the `compdef` function, there's no need to dig it deeper cause the info I told you is enough for a user to understand how completion works.)

Let find out what is `zcompint` and `zpcdreplay`.

```zsh
❯ which -a zpcompinit
zpcompinit () {
        autoload compinit
        compinit -d ${ZPLGM[ZCOMPDUMP_PATH]:-${ZDOTDIR:-$HOME}/.zcompdump} "${(Q@)${(z@)ZPLGM[COMPINIT_OPTS]}}"
}

❯ which -a zpcdreplay
zpcdreplay () {
        -zplg-compdef-replay -q
}

❯ which -- "-zplg-compdef-replay"
-zplg-compdef-replay () {
        local quiet="$1"
        typeset -a pos
        if [[ "${+functions[compdef]}" = "0" ]]
        then
                print "Compinit isn't loaded, cannot do compdef replay"
                return 1
        fi
        local cdf
        for cdf in "${ZPLG_COMPDEF_REPLAY[@]}"
        do
                pos=("${(z)cdf}")
                [[ "${#pos[@]}" = "1" && -z "${pos[-1]}" ]] && continue
                pos=("${(Q)pos[@]}")
                [[ "$quiet" = "-q" ]] || print "Running compdef ${pos[*]}"
                compdef "${pos[@]}"
        done
        return 0
}
```

`zcompinit` is `zcompinit` with custom settings. 

For the 2) type of completion. If you declare a normal function as completion function with `compdef` before `compinit`, it's useless cause `compdef` is not available yet. zdharma, the author of zplugin, try to use custom functions `zpcompdef` and `zpcdreplay` to ease this requirement. If a completion function is defined with `zpcompdef` instead of zsh's `zcompdef`, the completion function is saved to be parsed later. `zpcdreplay` is the time when these saved functions are parsed as completion functions. 

In fact, the `completion > fasd` (it should be `compinit > fasd`) part of the loading sequence is related to completion definition for `fasd`, `fasd`'s completion is defined by `compdef`. And the completions for `fasd` are hard to be separated to be declared by `zpcompdef`.

5. `zsh-users/zsh-completions` belongs to the 1) type of completion above. The plugin `zsh-users/zsh-completions` defines the completions in separated files, and add the directory containing these files into `fpath`. `ZPT::modules/completion` adds some additional settings and does `compinit` for you.

I'm tired after typing all these words. Dude, debug the loading sequence yourself in question No.5.

BTW: I haven't checked what I typed. Apology for any typos and semantic errors.

**Comment by ngocphamm** [comment]
Thanks again for a very detailed post @laggardkernel. I can't say I understand all of that, but definitely have had more insights into this whole thing.

From your recommendation, I think you are right that it's just the loading sequence of `fasd`. I moved it to after the `PZT::module/completion` and everything is fine now. I did some more Turbo for some other modules/plugins and now `time zsh -i -c echo` results in only about 50ms. 🎉 I can't be happier!

One thing about the `compdef` and Zplugin is that I think it does get captured by Zplugin for later parse with `zpcdreplay` because I have this line `compdef '_dispatch git git' config` defined before `zplugin ice wait"0" atinit"zpcompinit; zpcdreplay" lucid; zplugin light zdharma/fast-syntax-highlighting`, and at the prompt, using `zpl report` I can see that Zplugin captured that and saved for later.

Anyway, I don't quite plan on digging really deep into how everything in ZSH works because I don't work with shell script almost ever. At least for now. Still, your input is much appreciated!
